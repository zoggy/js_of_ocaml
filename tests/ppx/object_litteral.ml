(* Test methods. *)

fun (obj : < > Js.t) -> obj = object%js val m = () end
;;

fun (obj : < m : float Js.prop > Js.t) -> obj = object%js val mutable m = 0 end
;;

fun () ->
  (object%js
    val r = 2
    val mutable w = 3.
    method m = ""
  end : < m : int Js.meth; .. > Js.t)
;;


fun () ->
  object%js
    val r = 2
    val _r = 2
  end
;;

fun () ->
  (object%js
    val _r_a = 2
    val _r_b = 2
  end : <_r_a : int Js.readonly_prop > Js.t)
;;

fun () ->
  object%js
    val mutable w = 2
    val mutable _w = 2
  end
;;

fun () ->
  (object%js
    val mutable _w_a = 2
    val mutable _w_b = 2
  end : <_r_a : int Js.prop > Js.t)
;;

fun () ->
  object%js
    method m = ""
    method _m = ""
  end
;;

fun () ->
  (object%js
    method _m_a = ""
    method _m_b = ""
  end : <_m_a : string Js.meth; .. > Js.t)
;;

