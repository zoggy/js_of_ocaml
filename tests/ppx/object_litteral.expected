Characters 51-75:
  fun (obj : < > Js.t) -> obj = object%js val m = () end
                                ^^^^^^^^^^^^^^^^^^^^^^^^
Error: This expression has type < m : unit Js.readonly_prop > Js.t
       but an expression was expected of type <  > Js.t
       The second object type has no method m
Characters 49-80:
  fun (obj : < m : float Js.prop > Js.t) -> obj = object%js val mutable m = 0 end
                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: This expression has type < m : int Js.prop > Js.t
       but an expression was expected of type < m : float Js.prop > Js.t
       Type int Js.prop = < get : int; set : int -> unit > Js.gen_prop
       is not compatible with type
         float Js.prop = < get : float; set : float -> unit > Js.gen_prop 
       Types for method get are incompatible
Characters 14-84:
  ...object%js
      val r = 2
      val mutable w = 3.
      method m = ""
    end................................
Error: This expression has type
         < m : string Js.meth; r : int Js.readonly_prop; w : float Js.prop >
         Js.t
       but an expression was expected of type < m : int Js.meth; .. > Js.t
       Types for method m are incompatible
Characters 46-48:
      val _r = 2
          ^^
Duplicated val or method "_r" (normalized to "r").
Characters 32-33:
      val r = 2
          ^
Duplicated val or method "r".
Characters 49-53:
      val _r_b = 2
          ^^^^
Duplicated val or method "_r_b" (normalized to "r").
Characters 32-36:
      val _r_a = 2
          ^^^^
Duplicated val or method "_r_a" (normalized to "r").
Characters 61-63:
      val mutable _w = 2
                  ^^
Duplicated val or method "_w" (normalized to "w").
Characters 39-40:
      val mutable w = 2
                  ^
Duplicated val or method "w".
Characters 65-69:
      val mutable _w_b = 2
                  ^^^^
Duplicated val or method "_w_b" (normalized to "w").
Characters 40-44:
      val mutable _w_a = 2
                  ^^^^
Duplicated val or method "_w_a" (normalized to "w").
Characters 52-54:
      method _m = ""
             ^^
Duplicated val or method "_m" (normalized to "m").
Characters 34-35:
      method m = ""
             ^
Duplicated val or method "m".
Characters 56-60:
      method _m_b = ""
             ^^^^
Duplicated val or method "_m_b" (normalized to "m").
Characters 35-39:
      method _m_a = ""
             ^^^^
Duplicated val or method "_m_a" (normalized to "m").

